raiden/transfer/channel.py

```diff
@@ -887,11 +910,13 @@ def is_valid_withdraw_request(
 
 
 def is_valid_withdraw_confirmation(
-    channel_state: NettingChannelState, withdraw: ReceiveWithdraw
+    channel_state: NettingChannelState, withdraw: ReceiveWithdrawConfirmation
 ) -> SuccessOrError:
 
     result: SuccessOrError
 
+    withdraw_exists = withdraw.total_withdraw in channel_state.our_state.withdraws
+
     expected_nonce = get_next_nonce(channel_state.partner_state)
 
     packed = pack_withdraw(
```

Comment:
The withdraw *MUST* have the expiration as part of the signature. Otherwise this attack is possible:

Alice has balance 50
Bob has balance 0
Alice requests withdraw of 50
Bob sends confirmation
Alice waits for the withdraw to expire
Alice sends the withdraw expired message, Bob accepts
Alice sends a payment through Bob of value 50
After the payment is successful Alice uses the received withdraw confirmation to withdraw on-chain, effectively double spending.

Reply:
For that `raiden-contracts` needs to change https://github.com/raiden-network/raiden-contracts/issues/1102

Reply:
Done